### ISSUE TITLE
Metronome instability during tempo changes

### ISSUE THREAD

**Issue by brunchboy** [overtone/overtone#416]
While working on [porting](https://github.com/brunchboy/electro) my [enhanced version](
https://github.com/brunchboy/afterglow/blob/master/src/afterglow/rhythm.clj) of the metronome code to Java so I can use its features in [Beat Link](
https://github.com/brunchboy/beat-link#beat-link), my Java library for interacting with Pioneer CDJs, I finally figured out why the metronome feels unstable during rapid tempo changes. When I am using [Afterglow](
https://github.com/brunchboy/afterglow#afterglow) to run light shows, I have the metronome tempo tied to a rotary encoder on my Push 2, and I’m usually displaying a detailed visualization of the current beat, bar, and phrase position on the graphical display. Sometimes when I was speeding up the tempo quickly, I noticed the metronome seeming to jump to the wrong place. And now I understand why this happens.

The [code](https://github.com/overtone/overtone/blob/master/src/overtone/music/rhythm.clj#L107-L119) to reposition the metronome start point to keep you in the same musical place after a tempo change uses a seemingly nice finesse to keep you at the same point in the current beat by moving you only in multiples of the old and new beat duration, leaving the fractional part alone. Unfortunately, that doesn’t actually work. Consider the case where you are currently in the second half of a beat, and the new tempo is twice (or more) the old tempo; this means the “fractional part” we are leaving in place is actually now more than a full beat, and the metronome lands one beat further ahead of where it should have.

Thankfully, this is easy enough to address using the beat-phase code that I contributed a while back, but doing that carelessly can lead to a new problem: since you need to make two calculations about the metronome—to find the current beat, and the current beat phase—you need to perform both of those with respect to the same instant in time, or you might drift from one beat to the next in between asking the two questions, which would end up jumping you back to the beginning of the beat after the tempo change.

It was to address exactly this kind of issue that I implemented the “snapshot” support in my version of the metronome code, to support multiple coherent calculations about a particular point in time, even while system time advances. So, my [fixed version of the code](https://github.com/brunchboy/afterglow/blob/master/src/afterglow/rhythm.clj#L389-L398) can’t be copied in directly, as it relies on other pieces that don’t exist in the Overtone version. It also uses a simplified approach of maintaining a single `start` origin of the timeline for computing both beat and bar numbers (and I also support phrase counting and phrase phases). So the fix would need to be rewritten slightly to work in Overtone. This could be done without bringing over my snapshot facility in its entirety, but perhaps it is time to think about adopting more of my metronome enhancements at the same time?

**Comment by hlolli** [comment]
My answer is simple, yes. Taken into account that I've little understanding how the metronome was initially implemented and I use my own "metronome" algorithm when using Overtone. It's bit messy I think that the metronome is based on `System/currentTimeMillis` and a bpm change depends totally on how a musical pattern was made, since the metronome can only calculate future milliseconds with respect to the current state of the bpm value, seems to me. Instead the clock itself should be based on beat, makeing all bpm-changes present be valid in the future. That's like we both know, how abelton-link operates.

I'd be curious to learn, if you can elaborate better how you solved this problem. I'm still, ultra-slowly, trying to get abelton-link implemented in parallel to the at/at clock. But the at-at clock is maybe a child of its time and needs complete rethinking. A metronome using computer clock, tough accurate and high performing (a clock process already running in every computer) is probably not a direction we should head to. In ideal world we could just reimplement abelton-link algorithm, call it something else, communicate on the same port with abelton-link clients, but you would know better than me how the licensing from Abelton works in that regard?

**Comment by brunchboy** [comment]
I was able to fix this problem completely, and I think you underrate the Overtone metronome functions. 😄  They are a beautiful and extremely useful abstraction for working with musical time, so much so that when I reached the point in Beat Link where I was ready to start working on synchronizing CDJs with the Ableton Link timeline, I dropped all the protocol analysis and network coding I was working on, to port them to a pure Java implementation (linked in my first sentence in the issue—maybe looking at the [Java API docs](http://deepsymmetry.org/electro/apidocs/) will provide an alternate perspective on what I find so useful here).

`at-at` is something else completely, and I was pretty much stuck working on my lighting controller until I realized I should not try to schedule things, but instead should simply look at the time for which I am trying to generate a frame of lighting control values, and use the metronome to translate that time into a location within beats, bars, and phrases, to determine the proper lighting look. That uses the metronome only, not `at-at`; frames are generated at the appropriate rate for the hardware, much like audio is.

Let me try again to explain how I fixed `metro-bpm`. Its goal is to let you adjust the metronome configuration so that it is running at a different speed, but without moving you to a different point in musical time. The way it does that is set the tempo value, and then recompute when the metronome's first beat would have had to occur for you to still be in the same place. I just corrected that computation; formerly it was not properly handling the fractional part you had traveled through the current beat. The actual [commit](https://github.com/brunchboy/afterglow/commit/2c21f5365b7d9223a977bf969bfd3f5c37d5a84a#diff-16b9cd053eb5203e3e867cf2095b8c57) that implements that fix might make this more concrete, although as I mentioned above, the Overtone version of the metronome still has an extra bit of state tracking the start of the first bar, which I have eliminated in my simplification of the algorithm.

In essence, apart from when you are changing its configuration (and changing BPM counts as changing configuration), the Overtone metronome is a set of pure functions that let you map between clock time and musical time. As implemented in Overtone, you can only ask questions about the current clock time, so you can get inconsistent results when you ask about more than one piece of information, because you might have moved on to a new beat between calling the first and second functions. That’s why I added the snapshot mechanism. And the snapshot mechanism decouples the metronome from the computer clock. Even if you implement Ableton Link, you will want something like the Metronome functions to help work with beats, bars, phrases, phases, and the like.

As far as Link licensing… if you want to use their actual code, you have to use the GPL, which is incompatible with Clojure and Overtone. And they really, really discourage you trying to use your own code and interoperating with them. They can’t prevent that with licensing, but they are (not entirely unreasonably) worried about semi-broken implementations messing everything up. That’s why after talking things over with them I ended up implementing [Carabiner](
https://github.com/brunchboy/carabiner#carabiner) so I could use their implementation but control it over a network socket, to avoid licensing issues. You could do the same; here is the [code](https://github.com/brunchboy/beat-link-trigger/blob/master/src/beat_link_trigger/carabiner.clj) where I use it to adjust a Link session to track the tempo and beat grid of Pioneer CDJs. And if you keep an eye on the project, over the next week I expect to finish my implementation of offering synchronization in the other direction, with the help of my new Java port of the metronome functions.

**Comment by hlolli** [comment]
I think I'm getting you, and I looked bit more at your code and I think it shouldn't be a problem to import it to Overtone. Could you make a sound example, where you play say a scale or anything, and on some time interval, speed up and slow down the metronome, so I can compare. In my opinion, you should just make a PR and we continue from there, looks mostly like a copy paste from your code to me. If you make a minimal example which shows that this code improves the metronome, then it goes without saying that I'd merge it right away. The idea of the metronome is to quantize the ms value from the clock (now) and what you've done in your branch is continuation of that. I still don't understand how you sync the hardware framerate to the clock, because they surely could drift as far as I can see. Anyway, best would be if you could use at-at in your test example if you have time to do it.

**Comment by brunchboy** [comment]
Thanks, I will see if I have time to come up with such an example, but since I am already spending so much of my free (and sleep) time working on my own projects, I am not sure when I will find the time (especially since I will have to re-learn how to play music in Overtone, which I don’t normally do). Hmm, I just thought of one possible way to demonstrate the issue, so I may explore that.

But I have also been reflecting on the ways I use the metronome, and realize that most of the features I use heavily (the phase calculations and snapshot-based consistency) are ones I added myself, so perhaps I should just consider mine a fork, and leave things at that. I don’t know how heavily the metronome is even used in Overtone, and how common dynamic and abrupt tempo changes are, which would be the situations which would lead to this issue being a problem. (And using `at-at` to schedule notes has its own problems unrelated to the metronome or system clock, because Java thread scheduling provides no real-time guarantees; this [article](http://www.rationaljava.com/2015/10/measuring-microsecond-in-java.html?m=1) shows why I ended up deciding to tie up a CPU core busy-waiting on `System.nanotime()` in order to time my beat packets more precisely when taking over the sync master role with CDJs. However, a lag of a millisecond or two may not be deadly in an Overtone composition, especially since it would be jitter, not accumulating delay.

I was also reviewing the Ableton Link documentation and the [Linux Audio Conference paper](https://depositonce.tu-berlin.de/bitstream/11303/7886/4/LAC2018_proceedings.pdf#page=51) about it, and noticed how similar the Link timeline is to the approach I ended up with; it has the same notions of phase and capture-commit for consistent computations (what I called snapshots), along with the original metronome’s notion of mapping system time onto a musical timeline of beats and bars.

The reason I was looking in there was to try to better answer your question about coping with clock speed differences: I wanted to see if Link attempts to calculate any sort of scaling factor between system clocks. I also studied the Link code a bit (it is not the best-explained code I have ever encountered, but with the help of the paper it makes sense). As far as I can tell, they do not do any clock scaling; they assume, as do I, that the system clock is fairly good and can be relied on to authoritatively translate tempo and time to beat positions using the timeline linear equation. However, they periodically check to see if the timelines on different hosts are drifting apart, and if they are, they nudge them back into alignment (by shifting the timeline’s origin as needed), and doing this often enough that the shift is imperceptible. So you will want to set up a scheduled task to do the same thing between the Link timeline and the Overtone metronome (whether you are using my Carabiner or some other method to access the Link timeline). Does that make sense?

**Comment by brunchboy** [comment]
Although in case you are asking a slightly different question, about drift between my system clock and the hardware frame rate at which I am sending out DMX lighting frames, let me try to explain that as well: the beauty of this pure function approach is that there is absolutely no need to worry about synchronizing them. I determine when the next frame of lighting data will be needed, and obtain a metronome snapshot corresponding to that system time. That tells me what beat, bar, phrase (and/or fraction thereof) will be playing at the start of the frame, so I can use those values to generate the lighting look. I do this in advance to be sure it is ready when the hardware needs it. If I am not synchronizing with any external music player, that is the end of the story.

However, if I am synchronizing with (for example) a Pioneer CDJ network, then every now and then I will receive a beat packet which tells me I have drifted slightly out of synchronization, so I need to shift the metronome’s origin forward or back by a few milliseconds to get back in synch. If this happens before I generate a frame of lighting data, then it will be generated with the corrected synch. If it happens when I am already generating the lighting data, then (because I am using a snapshot), the old timeline position will be used for the whole frame, and it will be a few milliseconds out of synch, but it will be internally consistent. And then the next frame will be when we get back in synch. Since even at the fastest DMX frame rate, these are sent only every 25 milliseconds, these nudges will not move you past a beat boundary compared to the previous frame, and will be quite imperceptible.

**Comment by hlolli** [comment]
Sorry I've been to busy and short of coffee to fully read and digest your messages. I guess you could give me some of your energy :D 

Will give a reply soon as possible.

**Comment by brunchboy** [comment]
No rush… I think this can benefit from sitting on the back burner for a while as we both think about it. 😄
